base/reduce.jl

```diff
@@ -302,6 +302,10 @@ guaranteed left or right associativity and invocation of `f` for every value.
 mapreduce(f, op, itr; kw...) = mapfoldl(f, op, itr; kw...)
 mapreduce(f, op, itrs...; kw...) = reduce(op, Generator(f, itrs...); kw...)
 
+
+mapreduce(f::F, op::OP, x::Tuple) where {F,OP} = mapfoldl(identity, op, map(f, x))
```

Comment:
`map(f, x::Any32)` might has poor efficiency. Make them fallback to the general method?

Reply:
Also, `mapreduce(f, op, x::Tuple)` calls `mapfoldl(f, op, x::Tuple)` which should fallback to `Base.afoldl` anyway, thus would be fully unrolled if `length(x) < 32`. Why `foldl(op, map(f, x))` helps here?

Reply:
It helps sometimes, but not others.
I think I'll add a specialized `reduce` as well.
Maybe move `Static.reduce_tup` to Base:
https://github.com/SciML/Static.jl/blob/c75a720ad3cc7c32e95421144faa1be675c18fc4/src/Static.jl#L705
It's unrolled, type stable, and leads to reasonable SIMD orders for simpler types.

Reply:
The fun part will be trying to reimplement `reduce_tup` without the `@generated`.

Reply:
Unfortunately, `reduce_tup` would change the order of input:
```julia
julia> Static.reduce_tup(*, ("1","2","3","4"))
"1324"
```
So it's not legal for order sensitive reduction.

Reply:
I didn't test all cases, but examples from https://github.com/tpapp/TransformVariables.jl/pull/80 just caused by we dont call `afoldl` for `NamedTuple`. Adding the following seems a more general fix.
```julia
Base._foldl_impl(op, init, itr::NamedTuple) = Base._foldl_impl(op, init, Tuple(itr))
```

Reply:
From the doc string on `reduce`:

> The associativity of the reduction is implementation-dependent; if you need a particular
  associativity, e.g. left-to-right, you should write your own loop or consider using foldl or
  foldr. See documentation for reduce.

Reply:
Order is not defined/explicitly stated to be implementation-dependent.

Reply:
I'm not a expert on fold. But all `reduce` and `fold` in `Base` obeys the the same begin -> end rule (st present).
I think it would be confusing for users once we break it, as Iterator should be a ordered collection.

But of cource, many numerical code don't care it. Maybe we can add another keyword to turn on this performance optimization.

Reply:
It's always easier/faster to fix things in packages, so I'm fine continuing to just do that.

Reply:
Regarding the "order": Don't mix up associativity with commutativity: `reduce(*, ("1","2","3","4"))`is free to do `("1"*"2")*("3"*"4")` or `(("1"*"2")*"3")*"4"` or `"1"*("2"*("3"*"4"))` or ... (all of which give "1234") but not `"1"*"3"*"2"*"4"`.

Reply:
Ah, of course.